src/coreclr/jit/emitxarch.cpp

```diff
@@ -478,70 +478,150 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
 //    true if previous instruction zeroed reg's upper 32 bits.
 //    false if it did not, or if we can't safely determine.
 //
-// Notes:
-//    Currently only looks back one instruction.
-//
-//    movsx eax, ... might seem viable but we always encode this
-//    instruction with a 64 bit destination. See TakesRexWPrefix.
-
 bool emitter::AreUpper32BitsZero(regNumber reg)
 {
+    // If not a valid register, then return false.
+    if (!((reg >= REG_EAX) && (reg <= REG_FP_LAST)))
+        return false;
+
     // Only consider if safe
     //
     if (!emitCanPeepholeLastIns())
     {
         return false;
     }
 
-    instrDesc* id  = emitLastIns;
-    insFormat  fmt = id->idInsFmt();
+    bool result = false;
 
-    // This isn't meant to be a comprehensive check. Just look for what
-    // seems to be common.
-    switch (fmt)
-    {
-        case IF_RWR_CNS:
-        case IF_RRW_CNS:
-        case IF_RRW_SHF:
-        case IF_RWR_RRD:
-        case IF_RRW_RRD:
-        case IF_RWR_MRD:
-        case IF_RWR_SRD:
-        case IF_RWR_ARD:
+    emitPeepholeLastInstrs([&](instrDesc* id) {
+        switch (id->idInsFmt())
```

Comment:
If this is going to be depended on like this, we may need to consider auditing `instrsxarch.h` and `emitfmtsxarch.h`

There are some cases like `emitIns_R_R_I` where it's using `IF_RRW_RRW_CNS` but such an instruction doesn't exist. It should actually be either `IF_RWR_RRD_CNS` or `IF_RRW_RRD_CNS` depending on which instruction is being used.

Likewise, just glancing through `instrsxarch.h` there are many instructions which aren't correctly tracking `IUM`. For example, many of the `SIMD` instructions are listed as `IUM_WR` which while "technically correct" for the VEX encoding, will result in an incorrect format for the legacy encoding.

This is because many SIMD instructions are `RMW` in the legacy encoding but can encode an additional register for the VEX encoding.

So, for example, `addps` is encoded as `addps rwreg, rdreg/rdmem` with the legacy encoding, but as `addps wrreg, rdreg, rdreg/rdmem` for the VEX encoding.

For most SIMD instructions we go through `emitIns_SIMD_*` which will detect if we're doing the VEX encoding or not and select the right path to go down. For example, `emitIns_SIMD_R_R_R` will go to `emitIns_R_R_R` for `VEX` and emit an elideable `mov` + `emitIns_R_R` for the legacy path.

`emitIns_R_R_R` hardcodes the format to `IF_RWR_RRD_RRD` so it is already correct.
`emitIns_R_R` on the other hand uses `emitInsModeFormat(ins, IF_RRD_RRD)` which will then see `IUM_WR` which is incorrect as that will resolve `IF_RWR_RRD` when it should've resolved `IF_RRW_RRD`.

Just glancing through, it looks like most `binary` instructions are incorrect. As are the handful of `ternary` instructions.
Some `unary` instructions are correct, but others (like `movss`) don't account for the merge semantics that may exist.

There are also a handful of VEX encoded instructions (like `vfmaddsub132pd`) which are still RMW and which are `IF_RRW_RRD_RRD` (`ins rwreg, rdreg, rdreg/rdmem`) and that are also not tracking the `IUM` correctly

Reply:
We probably do need to vet these, but since `AreUpper32BitsZero` is only going to operate on GPRs, we don't have to worry too much about SIMD.

Also, I did add a few extra checks for instructions that output to two registers.